Add EIP-3664: Metaverse NFTs (#3664)
ethereum/EIPs: EIPS/eip-3664.md

---
eip: 3664
title: Next generation game NFT
author: TylerZ (@tylerztl), TY (@nimocat)
discussions-to: https://ethereum-magicians.org/t/eip-3664-next-generation-game-nft-standard/6738
status: Draft
type: Standards Track
category: ERC
created: 2021-07-17
requires: 20, 1155
---

> ## Simple Summary

Reviewer: Hi, this field was recently deprecated in favor of a `description` field in the preamble. Please see the [eip template](./eip-template.md) for proper formatting. Thanks!

Author: Thanks lightclient, already updated. Also updated the motivation. Thanks for your guide.

Author: @lightclient if you have any suggestions on our EIP, just tell us and we will make efforts on every aspects for pushing forward this EIP draft.

> This is a standard for next generation game NFTs, which extends ERC-1155.

Reviewer: ```suggestion
This is a standard for next generation game NFTs, which extends [ERC-1155](./eip-1155.md).
```
Please update your references to EIPs and ERCs to link to them.

Author: Thanks for your suggestions, we will working on details to satisfy the requirements, many thanks!

This EIP uses variants of ERC-20 represent the game attributes of NFTs. An ERC-3664 contract may have multiple attribute tokens. By limiting the transfer logic of these tokens, we can complete all game-related attributes through contracts. All kinds of NFT defined by ERC-3664 are gNFTs. If developers want to implement a complete RPG game prop system through smart contracts, they can use multiple gNFT contract combined with game logic contracts to develop the entire game in a decentralized form. This is an epic upgrade to the current blockchain game. ERC-3664 allows the props, characters and operation modes of the blockchain game to be fully represented in the form of smart contracts, which can be extended to a wide variety of different game play. ERC-3664 can also expand game NFT types on the Ethereum ecosystem, to bring Ethereum a stronger game ecosystem and various game play build on ERC-3664.

> ## Implementation

Reviewer: ```suggestion
## Specification
```

Should this be specification? You have `Implementation` twice.

---
eip: 3664

> title: ERC-3664 next generation game NFT

Reviewer: ```suggestion
title: Next generation game NFT
```

> author: TylerZ(@tylerztl), TY(@nimocat)

Reviewer: ```suggestion
author: TylerZ (@tylerztl), TY (@nimocat)
```

> discussions-to: <URL>

Reviewer: Please include a discussions URL. This is required before merging as a draft. We generally prefer people to create a thread on https://ethereum-magicians.org.

status: Draft
type: Standards Track
category: ERC
created: 2021-07-17
requires: 20, 1155
---

## Simple Summary

> ERC-3664 is a standard for next generation game NFTs, which extends ERC-3664.

Reviewer 2: I think there's a typo and it should be: "which extends ERC-1155"

Author: Thanks for remind, already changed.

> ERC-3664 is a standard for next generation game NFTs, which extends ERC-1155.

Reviewer: ```suggestion
This is a standard for next generation game NFTs, which extends ERC-1155.
```

Reviewer: The simple summary should explain what the EIP does, but this is just a longer version of the title.

## Abstract

> In this proposal, we created gNFT standard, which makes up for the problem that ERC-1155 can only support packed transfer and unchangeable attributes of NFTs. gNFT can make the attributes in the game props have four characteristics: transferable, upgradeable, changeable and evoluable, by using ERC-3664 standard NFTs in game, players are able to complete almost all operations on props only through smart contracts.

Reviewer 3: Are you french? :)

```suggestion
In this proposal, we created gNFT standard, which makes up for the problem that ERC-1155 can only support packed transfer and unchangeable attributes of NFTs. gNFT can make the attributes in the game props have four characteristics: transferable, upgradeable, changeable and evolvable, by using ERC-3664 standard NFTs in game, players are able to complete almost all operations on props only through smart contracts.
```

Author: Ah no, Singaporean. :)

> In this proposal, we created gNFT standard, which makes up for the problem that ERC-1155 can only support packed transfer and unchangeable attributes of NFTs. gNFT can make the attributes in the game props have four characteristics: transferable, upgradeable, changeable and evolvable, by using ERC-3664 standard NFTs in game, players are able to complete almost all operations on props only through smart contracts.

Reviewer: ```suggestion
In this proposal, we created the gNFT standard, which makes up for the problem that ERC-1155 can only support packed transfer and unchangeable attributes of NFTs. gNFT can make the attributes in the game props have four characteristics: transferable, upgradeable, changeable and evolvable. By using ERC-3664 standard NFTs in game, players are able to complete almost all operations on props only through smart contracts.
```

## Motivation

> ERC-3664 stems from a new RPG online game we were developed. We found that ERC1155 could not support all operations for game props, such as upgrading, adding attributes, and so on. It is almost impossible for the player to query all the attributes of the NFT on the smart contract, because ERC1155 cannot support the rich properties and functions of the props. For this reason, we propose a new ERC standard: ERC-3664, which is fully compatible with ERC-1155 and can help developers define massive operations on game NFTs. 

Reviewer: ```suggestion
We found that ERC-1155 had limited support for game assets, such as upgrading, adding attributes, and so on. It is almost impossible for the player to query all the attributes of the NFT on the smart contract, because ERC-1155 does not support the rich properties and functions of the assets. For this reason, we propose a new ERC standard which is fully compatible with ERC-1155 and can help developers define massive operations on game NFTs. 
```

> ERC-3664 uses variants of ERC20 represent the game attributes of NFTs. An ERC-3664 contract may have multiple attribute tokens. By limiting the transfer logic of these tokens, we can complete all game-related attributes through contracts. All kinds of NFT defined by ERC-3664 are gNFTs. If developers want to implement a complete RPG game prop system through smart contracts, they can use multiple gNFT contract combined with game logic contracts to develop the entire game in a decentralized form. This is an epic upgrade to the current blockchain game. ERC-3664 allows the props, characters and operation modes of the blockchain game to be fully represented in the form of smart contracts, which can be extended to a wide variety of different gameplay. ERC-3664 can also expand game NFT types on the Ethereum ecosystem, to bring Ethereum a stronger game ecosystem and various gameplay build on ERC-3664.

Reviewer: ```suggestion
This EIP uses variants of ERC-20 represent the game attributes of NFTs. An ERC-3664 contract may have multiple attribute tokens. By limiting the transfer logic of these tokens, we can complete all game-related attributes through contracts. All kinds of NFT defined by ERC-3664 are gNFTs. If developers want to implement a complete RPG game prop system through smart contracts, they can use multiple gNFT contract combined with game logic contracts to develop the entire game in a decentralized form. This is an epic upgrade to the current blockchain game. ERC-3664 allows the props, characters and operation modes of the blockchain game to be fully represented in the form of smart contracts, which can be extended to a wide variety of different game play. ERC-3664 can also expand game NFT types on the Ethereum ecosystem, to bring Ethereum a stronger game ecosystem and various game play build on ERC-3664.
```

        uint256[] memory _block_intervals
    ) external;

    /**
     * @dev Attach the attribute to NFT.
     */
    function attach(uint256 _nftId, uint256 _attrId) external;

    /**
     * @dev Remove the attribute from NFT.
     */
    function remove(uint256 _nftId, uint256 _attrId) external;

    /**
     * @dev Evolutive the attribute of the NFT.
     */
    function evolutive(uint256 _nftId, uint256 _attrId, uint8 _level) external;

    /**
     * @dev Evolution has the probability to damage the NFT, if it is damaged, it can be repaired.
     */
    function repair(uint256 _nftId, uint256 _attrId) external;

    event EvolutiveAttributeCreated(string name, uint256 id);
    event EvolutiveAttributeAttached(uint256 nftId, uint256 attrId);
    event EvolutiveAttributeRemoved(uint256 nftId, uint256 attrId);
    event AttributeEvolutive(uint256 nftId, uint256 attrId, uint8 level);
    event AttributeRepaired(uint256 nftId, uint256 attrId);
}

```

## Test Cases
[gNFT/test at main · DRepublic-io/gNFT (github.com)](https://github.com/DRepublic-io/gNFT/tree/main/test)

##  Implementation
[DRepublic-io/gNFT (github.com)](https://github.com/DRepublic-io/gNFT/)

> ## Additional References

Reviewer: Please remove this section, it isn't a valid section per EIP-1.

Author: Thank you lightclient, edited! Did you mean Additional References section? Removed, thanks!

---
eip: 3664
title: Next generation metaverse NFT

> description: This is a standard for next generation metaverse NFTs, which extends ERC-1155.

Reviewer 4: ```suggestion
description: Extend Non-Fungible Tokens with additional metadata for the metaverse
```

I think it can be inferred that this EIP is a standard, and "next generation" doesn't convey anything specific.

Reviewer 5: ```suggestion
description: Extend EIP-1155 Non-Fungible Tokens with additional metadata for the metaverse
```

@SamWilsn adding mentioning of EIP-1155 because this EIP seems only touch EIP-1155 not EIP-721 or other ERCs of NFTs.

author: TylerZ (@tylerztl), TY (@nimocat)
discussions-to: https://ethereum-magicians.org/t/eip-3664-next-generation-game-nft-standard/6738
status: Draft
type: Standards Track
category: ERC
created: 2021-07-17

> requires: 20, 1155

Reviewer 4: ```suggestion
requires: 20, 165, 1155
```

---

## Abstract

> In this proposal, we created the next gen component-NFT standard, which makes up for the problem that [ERC-721](./eip-721.md) and [ERC-1155](./eip-1155.md) can only support unchangeable attributes of NFTs. component-NFT can make the attributes in a NFT have five characteristics: transferable, upgradeable, changeable, combinable and evolvable. By using this standard NFTs in any fields, users can create any valuable dapps based on component-NFT standard.

Reviewer 4: The abstract should contain a _technical_ summary of the EIP, so readers can get a quick idea of what changes the EIP proposes. Should probably mention something about "attributes" becoming their own entity, that are attached to NFTs?

In our current proposal, we suggest a new type of extendable NFT metadata protocol, which is compatible with existing ERC721 and ERC1155, and can build on their basis to provide better descriptive power for attributes metadata.

Currently we provide the following options for attribute extendability:
1.	General attributes: for describing immutable attributes such as avatar birth dates.
2.	Variable attributes: for describing attributes whose value will change, such as an avatar’s combat power.
3.	Transferable attributes: for describing attributes that can be transferred to other NFTs.
4.	Upgradable attributes: for describing NFT levels and can trigger upgrades.
5.	Evolvable attributes: for describing the fact that NFT can evolve, and evolutions can fail, and if the case of failure, the NFT can no longer be used until properly repaired.

> 6.	Text attributes: for realizing functionality similar to Loot  since the above attributes are numerical in nature. 

Reviewer 4: This is less motivation, and more specification I think?

Reviewer 6: Rationale maybe?  Describing *why* each field is useful is a rationale thing.  Listing standard fields is a specification thing.  Explaining why the EIP in general is useful is a motivation thing.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
Every ERC-3664 compliant contract must implement the ERC3664 and ERC165 interfaces.

```solidity

> // SPDX-License-Identifier: MIT

Reviewer 4: Should be `CC0-1.0` to match the EIP itself.

pragma solidity ^0.8.0;

/**
 * @dev Required interface of an ERC3664 compliant contract.
 */
interface ERC3664 {
    /**
     * @dev Emitted when new attribute type `attrId` are minted.
     */
    event AttributeCreated(
        uint256 indexed attrId,
        string name,
        string symbol,
        string uri
    );

    /**
     * @dev Emitted when `value` of attribute type `attrId` are attached to "to"
     * or removed from `from` by `operator`.
     */
    event TransferSingle(
        address indexed operator,
        uint256 from,
        uint256 to,
        uint256 indexed attrId,
        uint256 value
    );

    /**
     * @dev Equivalent to multiple {TransferSingle} events.
     */
    event TransferBatch(
        address indexed operator,
        uint256 from,
        uint256 to,
        uint256[] indexed attrIds,
        uint256[] values
    );

    /**

>      * @dev Returns primary attribute type of owned by `tokenId`.

Reviewer 4: This reads weirdly, so I'm not sure what it means. Would you mind rewording it?

        uint256 amount
    ) external;

    /**
     * @dev [Batched] version of {attach}.
     */
    function batchAttach(
        uint256 tokenId,
        uint256[] calldata attrIds,
        uint256[] calldata amounts
    ) external;
}

interface ERC165 {
    /// @notice Query if a contract implements an interface
    /// @param interfaceID The interface identifier, as specified in ERC-165
    /// @dev Interface identification is specified in ERC-165. This function
    ///  uses less than 30,000 gas.
    /// @return `true` if the contract implements `interfaceID` and
    ///  `interfaceID` is not 0xffffffff, `false` otherwise
    function supportsInterface(bytes4 interfaceID) external view returns (bool);
}
```

The first step to apply ERC3664 is to define one or more attributes with _mint or _mintBatch. An NFT creator could attach or batchAttach from a predetermined attribute list to make their NFT. This operation can be completed during the creation of the NFT or may be dynamically applied to update NFT attributes. Hence we have redefined the uniqueness of NFT. It is not just the unique tokenId , but also the list of attributes that differentiate an NFT from others.

> ## `View` Functions

Reviewer: Not a valid EIP section per EIP-1.

### Evolvable Attribute

```solidity
/**
 * @dev Interface for the Evolvable functions from the ERC3664.
 */
interface IERC3664Evolvable is IERC3664Metadata {
    event AttributeEvolvable(
        address indexed operator,
        uint256 tokenId,
        uint256 attrId,
        bool status
    );

    event AttributeRepaired(
        address indexed operator,
        uint256 tokenId,
        uint256 attrId
    );

    function period(uint256 tokenId, uint256 attrId)
        external
        view
        returns (uint256);

    function evolutive(uint256 tokenId, uint256 attrId) external;

    function repair(uint256 tokenId, uint256 attrId) external;
}
```

Evolvable attributes describe how NFTs cans strengthen attributes via evolution, which yields uncertain success with different probabilities. NFT creators can set the probability of success of each evolution. NFTs that fail to evolve become useless. They can only be used and evolved normally after proper repairs. This feature provides templates for evolutionary functions for the metaverse NFTs, which can be modified as needed for other projects that require evolutionary approaches.

> ## Rational

Reviewer: ```suggestion
## Rationale
```

ERC3664 itself does not create any NFTs, and is only used for storing and expressing NFT attributes. Therefore it is a completely new NFT attribute extendability standard that seeks to tokenize attributes. Every attribute has its own independent ID, and can be independently transferred and combined to create richly diverse digital artworks with NFTs, and not just fixed collectibles.

> ## Compatibility

Reviewer: ```suggestion
## Compatibility
```
Not a valid EIP section per EIP-1.

ERC3664 is compatible to any kind of NFT protocol of any standard. It functions merely as an NFT supplementary protocol rather than as a replacement protocol. Hence most NFT works in the markets do not need to be modified and need to realize ERC3664 by appending attributes the existing NFT token. The original work can be displayed through ERC3664’s tokenURI. Support for ERC721 and ERC1155 is non-intrusive.

> ## Use Cases

Reviewer: Not valid EIP section per EIP-1.

[Metacore](https://github.com/DRepublic-io/drepublic-contracts/blob/main/contracts/Metacore.sol)

## Reference Implementation
https://github.com/DRepublic-io/drepublic-contracts

## Security Considerations
Since the owner becomes vacuous after every ERC-721 is aggregated by ERC3664, and since each contract only allows the contract owner to invoke functions in contract，ERC-721 and ERC-1155 aggregated by EIP-3664 are guaranteed to be safe.

## Copyright

> Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

Reviewer 7: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```

The NFT standard proposed by EIP-3664 enable unlimited disaggregation and assembly. Players can forge modularized NFTs with this proposal, and aggregate modularized NFTs into a new NFT, and let new NFTs combine with each other. On the other hand, NFTs can be disaggregated into component parts and reassembled. These powerful functionalities can sustain many different use cases in the future, such as team configuration in competitive gaming, amendable terms and clauses in financial contracts and honor awarding mechanisms in blockchain games. With the present proposal, you can realize any arbitrary functionality with your NFT. At the same time, the present proposal is an extension protocol of ERC-1155 and ERC-721, implying compatibility with ERC-721 and ERC-1155, and enabling the aggregation of NFTs on old protocols into newly modularized ones. 

## Specification

ERC-3664 extends ERC-1155, divides NFT attributes into five categories:

> 1. General changeable attributes: usually refers to basic attributes such as ATK and DEF. Its functions include increase or decrease in attribute value.

Reviewer 8: What is ATK and DEF?

Author: ATK and DEF are a Attack and Defense attributes which are often used in games. EIP-3664 are really suitable for implementing game props based on layer 2 solutions, also quite useful for DeFi tools. I will remove some game related terms in order to make it more general to readers. Thanks

2. Transferable attributes: refers to the NFT's transferable attributes can be transferred to other NFTs in the event of destruction or other circumstances.
3. Upgradeable attributes: refers to the NFT that can be upgraded according to the formula, the player pays the fee, and it is reflected in the increase of the level, usually accompanied by the increase of other attribute values.
4. Combinable, NFTs can be combined with other NFTs, to make all NFT world combinable and separatable, each NFT is a component of other NFTs, freely combine to make greater value.
5. Evolvable attribute: refers to the attribute that NFT may evolve automatically over time (the height of the chain block), simulate the time attributes in the real world, so that NFT has the characteristics of time-varying.

```solidity

> function remove(uint256 _nftId, uint256 _attrId) external;

Reviewer 8: I think this EIP could be structured better. I am not sure what these functions do and how the interfaces interact with them.

Many EIPs have chosen to describe the interface as a natspec documented Solidity interface. While I think we should not tie specification to Solidity, not better alternative have shown up yet.

Author: Thanks for your suggestions, we will convert the format the natspec standard interface asap.

```

Remove the attribute from NFT. only called by contract operator.

- Param1: attribute owned id 
- Param2: attribute id

### IGenericAttribute

> ```

Reviewer 8: ```suggestion
```solidity
```

    /**
     * @dev Evolution has the probability to damage the NFT, if it is damaged, it can be repaired.
     */
    function repair(uint256 _nftId, uint256 _attrId) external;

    event EvolutiveAttributeCreated(string name, uint256 id);
    event EvolutiveAttributeAttached(uint256 nftId, uint256 attrId);
    event EvolutiveAttributeRemoved(uint256 nftId, uint256 attrId);
    event AttributeEvolutive(uint256 nftId, uint256 attrId, uint8 level);
    event AttributeRepaired(uint256 nftId, uint256 attrId);
}

```

`evolutive` function

```solidity
function evolutive(uint256 _nftId, uint256 _attrId, uint8 _level) external;
```

Evolutive the attribute of the NFT.  only called by nft creator.

- Param1: attribute owned id
- Param2: attribute recipient id
- Param3: attribute level

`repair` function

```solidity
function repair(uint256 _nftId, uint256 _attrId) external;
```

Evolution has the probability to damage the NFT, if it is damaged, it can be repaired. only called by nft creator.

- Param1: attribute owned id
- Param2: attribute recipient id
- Param3: attribute sub level

## Backward Compatibility

> The EIP-3664 standard can plug into existing ERC-721 and ERC-1155 interfaces (接口), and regard them as part of the modularized NFT, which uses EIP-3664 to extend and aggregate.

Reviewer 8: ```suggestion
The EIP-3664 standard can plug into existing ERC-721 and ERC-1155 interfaces, and regard them as part of the modularized NFT, which uses EIP-3664 to extend and aggregate.
```

---
eip: 3664
title: Next generation metaverse NFT

> description: This is a standard for next generation metaverse NFTs, which extends [ERC-1155](./eip-1155.md)

Reviewer: ```suggestion
description: Standard for next-generation metaverse NFTs, which extend ERC-1155
```

Reviewer: @axic do you know if links are okay in the description field? My guess is no, but not sure.

Reviewer 8: Sorry I missed this. Probably not as we run the escaping on it. The same applies to `withdrawal-reason` where I saw it is not rendered properly.

I think it would be nice to render both, but for the sake of merging this I'd consider it not possible.

Reviewer: @nimocat still need to get this fixed before merging.

author: TylerZ (@tylerztl), TY (@nimocat)
discussions-to: https://ethereum-magicians.org/t/eip-3664-next-generation-game-nft-standard/6738
status: Draft
type: Standards Track
category: ERC
created: 2021-07-17
requires: 20, 1155
---

## Abstract

> In this proposal, we created the next gen component-NFT standard, which makes up for the problem that ERC-721 and ERC-1155 can only support unchangeable attributes of NFTs. component-NFT can make the attributes in a NFT have five characteristics: transferable, upgradeable, changeable, combinable and evolvable. By using ERC-3664 standard NFTs in any fields, users can create any valueable dapps based on component-NFT standard.

Reviewer: ```suggestion
In this proposal, we created the next gen component-NFT standard, which makes up for the problem that [ERC-721](./eip-721.md) and [ERC-1155](./eip-1155.md) can only support unchangeable attributes of NFTs. component-NFT can make the attributes in a NFT have five characteristics: transferable, upgradeable, changeable, combinable and evolvable. By using this standard NFTs in any fields, users can create any valuable dapps based on component-NFT standard.
```
This last sentence seems a bit confusing?

## Motivation

> In the early days of NF, ERC-721 was the main NFT protocol. It focused on differentiating NFTs from FTs. But since the emergence of the Layer 2 and Metaverse, outdated NFT protocols would no longer suffice for NFT standards in the new era. We hope to objectize NFTs, and allow for their upgrades, assembly and disaggregation on the blockchain. We will also let them change with the passage of time so that these fundamental functionalities will endow NFTs with a world of possibilities, creating unparalleled value.

Reviewer: ```suggestion
In the early days of NFTs, ERC-721 was the main NFT protocol. It focused on differentiating NFTs from FTs. But since the emergence of the Layer 2 and Metaverse, outdated NFT protocols would no longer suffice for NFT standards in the new era. We hope to objectize NFTs, and allow for their upgrades, assembly and disaggregation on the blockchain. We will also let them change with the passage of time so that these fundamental functionalities will endow NFTs with a world of possibilities, creating unparalleled value.
```
Spelling error?

Reviewer: Also note, ERC-721 isn't a protocol. It's a standard.

## Specification

ERC-3664 extends ERC-1155, divides NFT attributes into five categories:

1. General changeable attributes: usually refers to basic attributes such as ATK and DEF. Its functions include increase or decrease in attribute value.
2. Transferable attributes: refers to the NFT's transferable attributes can be transferred to other NFTs in the event of destruction or other circumstances.
3. Upgradeable attributes: refers to the NFT that can be upgraded according to the formula, the player pays the fee, and it is reflected in the increase of the level, usually accompanied by the increase of other attribute values.
4. Combinable, NFTs can be combined with other NFTs, to make all NFT world combinable and separatable, each NFT is a component of other NFTs, freely combine to make greater value.
5. Evolvable attribute: refers to the attribute that NFT may evolve automatically over time (the height of the chain block), simulate the time attributes in the real world, so that NFT has the characteristics of time-varying.

```solidity
function remove(uint256 _nftId, uint256 _attrId) external;
```

Remove the attribute from NFT. only called by contract operator.

> - Param1：attribute owned id 

Reviewer: ```suggestion
- Param1: attribute owned id 
```

> - Param2：attribute id

Reviewer: ```suggestion
- Param2: attribute id
```

    event EvolutiveAttributeAttached(uint256 nftId, uint256 attrId);
    event EvolutiveAttributeRemoved(uint256 nftId, uint256 attrId);
    event AttributeEvolutive(uint256 nftId, uint256 attrId, uint8 level);
    event AttributeRepaired(uint256 nftId, uint256 attrId);
}

```

`evolutive` function

```solidity
function evolutive(uint256 _nftId, uint256 _attrId, uint8 _level) external;
```

Evolutive the attribute of the NFT.  only called by nft creator.

- Param1：attribute owned id
- Param2：attribute recipient id
- Param3：attribute level

`repair` function

```solidity
function repair(uint256 _nftId, uint256 _attrId) external;
```

Evolution has the probability to damage the NFT, if it is damaged, it can be repaired. only called by nft creator.

- Param1：attribute owned id
- Param2：attribute recipient id
- Param3：attribute sub level

## Backward Compatibility
The EIP-3664 standard can plug into existing ERC-721 and ERC-1155 interfaces (接口), and regard them as part of the modularized NFT, which uses EIP-3664 to extend and aggregate.

Thus any contract that has already deployed ERC-721 and ERC-1155 could be extended, assembled and disaggregated with EIP-3664.

## Test Cases

> 

Reviewer: ```suggestion
```

}
```

Every attribute is an independent component of an NFT. Transferrable attributes can be traded based on its protocol. After the NFT beneficiary receives the new attribute, the original NFT ID remains unchanged, but its price and rarity will increase.

### Updatable Attribite

```solidity
/**
 * @dev Interface for the updatable functions from the ERC3664.
 */
interface IERC3664Updatable is IERC3664Metadata {
    /**
     * @dev Remove attribute type `attrId` from `tokenId`.
     */
    function remove(uint256 tokenId, uint256 attrId) external;

    /**
     * @dev Increases `amount` value of attribute type `attrId` to `tokenId`.
     */
    function increase(
        uint256 tokenId,
        uint256 attrId,
        uint256 amount
    ) external;

    /**
     * @dev Decreases `amount` value of attribute type `attrId` from `tokenId`.
     */
    function decrease(
        uint256 tokenId,
        uint256 attrId,
        uint256 amount
    ) external;
}
```

Updatable attributes of NFTs can change under certain trigger conditions. Such use cases are common in gaming and finance. For example, the life force of an character, combative power of weaponry, debt interest rates, etc.

> 

Reviewer 9: "the life force of `an` character" - there is a grammar mistake... change an to a
